### PR TITLE
Enable FSDP2 cpu offloading

### LINKED ISSUE #620: Is there way to offload training memory to DRAM (using FSDP2?) for training Llama3-8B with torchtitan?
I am training Llama3-8B using 2 RTX A6000ada 48GB, but got OOM. Is there way to offload training memory to DRAM (using FSDP2?) for training Llama3-8B with torchtitan?

Thanks!

***Error message:
torch.OutOfMemoryError: CUDA out of memory. Tried to allocate 112.00 MiB. GPU 0 has a total capacity of 47.48 GiB of which 92.81 MiB is free. Including non-PyTorch memory, this process has 46.71 GiB memory in use. Of the allocated memory 45.56 GiB is allocated by PyTorch, and 448.27 MiB is reserved by PyTorch but unallocated. If reserved but unallocated memory is large try setting PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True to avoid fragmentation.  See documentation for Memory Management  (https://pytorch.org/docs/stable/notes/cuda.html#environment-variables)

***Here is my training config:

# torchtitan Config.toml
# NOTE: this toml config is a preset for 64 A100 GPUs.

[job]
dump_folder = "./outputs"
description = "Llama 3 8B training"

[profiling]
enable_profiling = true
save_traces_folder = "profile_trace"
profile_freq = 100

[metrics]
log_freq = 10
enable_tensorboard = true
save_tb_folder = "tb"

[model]
name = "llama3"
flavor = "8B"
norm_type = "rmsnorm"  # layernorm / np_layernorm / rmsnorm / fused_rmsnorm
tokenizer_path = "./torchtitan/datasets/tokenizer/original/tokenizer.model"

[optimizer]
name = "AdamW"
lr = 3e-4

[training]
batch_size = 2 #1
seq_len = 256 #512 #8192
warmup_steps = 200  # lr scheduler warm up
max_norm = 1.0  # grad norm clipping
steps = 1000
data_parallel_replicate_degree = 1  #1
data_parallel_shard_degree = -1 #-1
tensor_parallel_degree = 2 #1
compile = true
dataset = "c4"

[experimental]
pipeline_parallel_degree = 1 #1
enable_async_tensor_parallel = true

[checkpoint]
enable_checkpoint = false #false
folder = "checkpoint"
interval_type = "steps"
interval = 500
model_weights_only = false
export_dtype = "bfloat16"  #32
async_mode = "disabled" # ["disabled", "async", "async_with_pinned_mem"]

[activation_checkpoint]
mode = 'selective'  # ['none', 'selective', 'full']
selective_ac_option = 'op'  # 'int' = ac every positive int layer or 'op', ac based on ops policy

[float8]
enable_float8_linear = true
enable_fsdp_float8_all_gather = true
precompute_float8_dynamic_scale_for_fsdp = true

### PR DESCRIPTION
**Pull request by mori360**
resolve #620 
Add config: `--training.enable_cpu_offload`

Command: `CONFIG_FILE="./train_configs/llama3_8b.toml" ./run_llama_train.sh`

For non-pp case:
Screenshot 2024-10-23 at 1 45 56 PM

For pp case:
cpu offload+pp

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -169,13 +166,15 @@ def loss_fn(pred, labels):
             else "cuda"
         )
         model.to_empty(device=init_device)
-        model.init_weights()
+        if job_config.training.enable_cpu_offload:
+            with torch.device("cuda"):
```
Comment (tianyu-l) [inline train.py]: I somehow think the `model.init_weights(buffer_device="cuda")` change sounds better. It is straightforward on what we want to achieve, while making minimum change to code.

File: test_runner.py [hunk 2]
```diff
@@ -57,10 +57,12 @@ def build_test_list():
             [
                 [
                     "--training.compile",
+                    "--training.enable_cpu_offload True",
```
Comment (mori360) [inline test_runner.py]: Add into 1D compile test first

Comment (tianyu-l) [reply test_runner.py]: sorry I didn't make it clear during offline discussion.
I think we shouldn't touch this test. Instead we should create a new one with `enable_cpu_offload` only

File: torchtitan/config_manager.py [hunk 3]
```diff
@@ -253,6 +253,15 @@ def __init__(self):
...
             1 means disabled.""",
         )
+        self.parser.add_argument(
+            "--training.enable_cpu_offload",
+            type=bool,
+            default=False,
+            help="""
+            The `enable_cpu_offload` argument specifies whether to have offloading policy
+            for FSDP. If True, CPU offload of parameters, gradients, and optimizer states
+            will be supported.""",
```
Comment (tianyu-l) [inline torchtitan/config_manager.py]:
```suggestion
            help="""
            Whether to apply CPU offloading of parameters, gradients, and optimizer states in FSDP""",
```

File: torchtitan/models/llama/model.py [hunk 4]
```diff
@@ -391,8 +394,12 @@ def init_weights(self):
...
         ``Transformer`` root module to avoid reinitializing tensors.
         """
-        with torch.device(self.freqs_cis.device):
-            self.freqs_cis = self._precompute_freqs_cis()
+        if buffer_device is not None:
+            with torch.device(buffer_device):
+                self.freqs_cis = self._precompute_freqs_cis()
+        else:
+            with torch.device(self.freqs_cis.device):
+                self.freqs_cis = self._precompute_freqs_cis()
```
Comment (tianyu-l) [inline torchtitan/models/llama/model.py]:
```suggestion
        buffer_device = buffer_device or self.freqs_cis.device:
        with torch.device(buffer_device):
            self.freqs_cis = self._precompute_freqs_cis()
```

File: train.py [hunk 5]
```diff
@@ -159,9 +159,15 @@ def loss_fn(pred, labels):
...
+        init_device = (
+            "cpu"
+            if job_config.checkpoint.create_seed_checkpoint
+            or job_config.training.enable_cpu_offload
+            else "cuda"
+        )
         model.to_empty(device=init_device)
-        model.init_weights()
+        buffer_device = "cuda" if job_config.training.enable_cpu_offload else None
+        model.init_weights(buffer_device=buffer_device)
```
Comment (tianyu-l) [inline train.py]: can we just do
```suggestion
        model.init_weights(buffer_device="cuda")
```

File: train.py [hunk 6]
```diff
@@ -159,9 +159,15 @@ def loss_fn(pred, labels):
         models_parallelize_fns[model_name](model, world_mesh, parallel_dims, job_config)
 
         # move sharded model to CPU/GPU and initialize weights via DTensor
-        init_device = "cpu" if job_config.checkpoint.create_seed_checkpoint else "cuda"
+        init_device = (
+            "cpu"
+            if job_config.checkpoint.create_seed_checkpoint
+            or job_config.training.enable_cpu_offload
```
Comment (tianyu-l) [inline train.py]: Don't we need to do the same for the PP case (several lines above)? Or are we assuming if PP is used, CPU offloading is not an option?

**Review by tianyu-l — COMMENTED** [review]
Looks good in general. Had some inline comments. In particular, let's figure out if CPU offloading and PP should coexist; if so we should add support for that as well.

File: train.py (hunk 5 above)
Comment (mori360) [reply train.py]: Not sure whether the change here would affect the case with:
```
--job_config.checkpoint.create_seed_checkpoint = True
--job_config.training.enable_cpu_offload = False
```

Comment (tianyu-l) [reply train.py]: good point

File: train.py (hunk 6 above)
Comment (mori360) [reply train.py]: I test it with pp.
Based on `llama2_7b.toml`, cpu_offload also works but has a much obvious latency in training.
Screenshot 2024-10-25 at 3 40 25 PM

File: test_runner.py [hunk 7]
```diff
@@ -351,6 +351,17 @@ def build_test_list():
...
+            [
+                [
+                    "--experimental.pipeline_parallel_degree 2",
+                    "--training.enable_cpu_offload True",
+                ],
+            ],
+            "Enable CPU Offload with PP",
+            "enable_cpu_offload+PP",
+            ngpu=4,
+        ),
```
Comment (mori360) [inline test_runner.py:364]: test with pp, could remove pp later if not necessary in the CI test

File: train.py [hunk 8]
```diff
@@ -151,17 +151,25 @@ def loss_fn(pred, labels):
         for m in model_parts:
             # apply SPMD-style PT-D techniques
             models_parallelize_fns[model_name](m, world_mesh, parallel_dims, job_config)
-            m.to_empty(device="cuda")
-            m.init_weights()
+            init_device = "cpu" if job_config.training.enable_cpu_offload else "cuda"
+            m.to_empty(device=init_device)
+            buffer_device = "cuda" if job_config.training.enable_cpu_offload else None
+            m.init_weights(buffer_device=buffer_device)
```
Comment (tianyu-l) [inline train.py]: Can we do one more change, to move `init_device` and `buffer_device` to be outside `if parallel_dims.pp_enabled`? This way the code can be shared by the two branches.

Also I think it reads clearer if we change them to something like
```suggestion
        if job_config.training.create_seed_checkpoint:
            init_device = "cpu"
            buffer_device = None
        elif job_config.training.enable_cpu_offload:
            init_device = "cpu"
            buffer_device = "cuda"
        else:
            init_device = "cuda"
            buffer_device = None
```

**Review by tianyu-l — APPROVED** [review]
lgtm! thanks!

**Pull request merged**